lib/Backends/NNPI/InferencePool.cpp

```diff
@@ -159,267 +227,304 @@ void InferenceThreadEnv::execute(RunIdentifierTy runId,
-          ERROR, hostOutputs_.size() == rawOutputs_.size(), "Bad outputs",
-          runId, ctx, resultCB);
-      if (deviceOptions_->enabledCommandLists > 1) {
-        uint32_t numErrors(0);
-        NNPIInferenceErrorCode result = nnpiCommandListWait(
-            commandList_, UINT32_MAX, commandErrors_, numCommands_, &numErrors);
-
-        if (result != NNPI_INF_NO_ERROR) {
-          LOG_NNPI_INF_ERROR(result, "Failed to wait on command list");
-        } else {
-          for (uint32_t i = 0; i < numErrors; i++) {
-            LOG(ERROR) << NNPI_INF_ERROR_MSG(commandErrors_[i].err,
-                                             commandErrors_[i].desc);
-          }
-        }
-        if (result != NNPI_INF_NO_ERROR || numErrors > 0) {
-          LOG_AND_CALLBACK_EXECUTE_NNPI_INF_ERROR(
-              nnpiCommandListClearErrors(commandList_),
-              "Failed to clear command list errors", runId, ctx, resultCB);
-          LOG_AND_FAIL_EXECUTE_CALLBACK_IF_NOT(
-              ERROR, false /* fail */, "Errors found in command list execution",
-              runId, ctx, resultCB);
+  // Post inference output handling and wait for outputs.
+  if (deviceOptions_->inferOnDevice) {
+    // For every output, lock and copy data (+convert), unlock
+    // then copy to output tensors.
+    LOG_AND_FAIL_EXECUTE_CALLBACK_IF_NOT(
+        ERROR, hostOutputs_.size() == rawOutputs_.size(), "Bad outputs", runId,
+        ctx, resultCB);
+    if (deviceOptions_->enabledCommandLists > 1) {
+      uint32_t numErrors(0);
+      NNPIInferenceErrorCode res = nnpiCommandListWait(
+          commandList_, UINT32_MAX, commandErrors_, numCommands_, &numErrors);
+      TRACE_EVENT_END(ctx->getTraceContext(), TraceLevel::OPERATOR,
+                      TRACING_INFERENCE);
+      TRACE_EVENT_BEGIN(ctx->getTraceContext(), TraceLevel::COPY,
+                        TRACING_POST_PROCESS);
+      if (res != NNPI_INF_NO_ERROR) {
+        LOG_NNPI_INF_ERROR(res, "Failed to wait on command list");
+      } else {
+        for (uint32_t i = 0; i < numErrors; i++) {
+          LOG(ERROR) << NNPI_INF_ERROR_MSG(commandErrors_[i].err,
+                                           commandErrors_[i].desc);
         }
       }
-      for (size_t i = 0, e = hostOutputs_.size(); i < e; i++) {
-        void *pHostOutput(hostOutputs_[i].hostPtr);
-
-        // Lock output host resource.
-        if (deviceOptions_->enabledCommandLists < 2) {
-          LOG_AND_CALLBACK_EXECUTE_NNPI_INF_ERROR(
-              nnpiHostResourceLock(hostOutputs_[i].handle, NNPI_LOCK_FOR_READ,
-                                   UINT32_MAX, &pHostOutput),
-              "Failed to lock host resource", runId, ctx, resultCB);
-          LOG_AND_FAIL_EXECUTE_CALLBACK_IF_NOT(ERROR, pHostOutput, "Bad output",
-                                               runId, ctx, resultCB);
-        }
-
-        // Copy data to output tensor.
+      if (res != NNPI_INF_NO_ERROR || numErrors > 0) {
```

Comment:
Yeah same as above -- so it's possible that we have `NNPI_INF_NO_ERROR` and also have errors which seems weird. Perhaps we can add some comments here to explain a bit.